src/python/dxpy/sugar/processing.py

```diff
@@ -0,0 +1,508 @@
+from __future__ import print_function, unicode_literals, division, absolute_import
+import copy
+import logging
+import shlex
+import subprocess
+import tempfile
+
+from dxpy import compat
+
+
+LOG = logging.getLogger()
+
+
+# stdout/stderr types
+_PIPE = 0
+_FILE = 1
+_BUFFER = 2
+_OTHER = 3
+
+
+class Processes:
+    """
+    Encapsulates one or more commands, runs those commands using the
+    `subprocess` module, and provides access to the results.
+
+    Args:
+        cmds (sequence): Command strings or sequences of command arguments to
+            be chained together.
+        stdout: How to capture stdout of the final process in the chain; can be
+            a string (filename) or file object, in which case the output is written
+            to that file; None, in which case `subprocess.PIPE` is used; True,
+            in which case output is captured to a buffer; or False, in which case
+            stdout is discarded. If a file, any existing file with the same name is
+            overwritten.
+        stderr: How to capture stderr of the final process in the chain (see
+            `stdout` for details).
+        capture_stderr: Whether to capture the contents of stderr of processes
+            other than the final process.
+        echo: Whether to echo commands to the logger before running the commands.
+            Can be overridden by the `run()` method's `echo` parameter.
+        popen_kwargs: Keyword arguments to pass to Popen constructors.
+
+    Todo:
+        Add ability to send input to stdin of first process.
+    """
+    def __init__(
+        self, cmds, stdout=None, stderr=None, capture_stderr=True, echo=None,
+        **popen_kwargs
+    ):
+        self.cmds = cmds
+        self._stdout = stdout
+        self._stdout_type = None
+        self._stderr = stderr
+        self._stderr_type = None
+        self.capture_stderr = capture_stderr
+        self._stderr_buffers = [] if capture_stderr else None
+        self._closed = False
+        self.echo = echo
+        self.popen_kwargs = popen_kwargs
+        self._processes = None
+        self._output_handle = None
+        self._returncode = None
+        # These will hold
+        self.out = None
+        self.err = None
+
+    @property
+    def returncode(self):
+        """
+        The max return code of all the commands in the chain.
+        """
+        if self._returncode is None:
+            returncodes = [
+                proc.returncode for proc in self._processes if proc.poll() is not None
+            ]
+            if returncodes:
+                # TODO: Is setting to max the right behavior?
```

Comment:
I think returning the **first** failed command's error code is the correct thing to do. Mimicing the behaviour of `set -o pipefail`.

Devil's advocating a bit. If we wish to depart from `set -o pipefail` behaviour in this aspect then we should return the **first** failed error code In practice when piping, later failures are usually due to earliers ones.

```
$ dx describe file-xxxx --json | grep "name"
# DXCLIError: No matches found for file-xxxx
```

The return code in the normal `set -o pipefail` case is from the failed `grep`, 1. The error occured with the `dx` command, exit code 3.


Again, to be clear, I vote **return the error code of the last command** to keep with expectations set by `set -o pipefail`. Still show the failure message of the real failed command.

Reply:
Done